### DISCUSSION TITLE
get chatid with php

### DISCUSSION THREAD

**Discussion by DeAref** [revenkroz/telegram-web-app-bot-example#5] [Q&A] [answered]
How to get Telegram.WebApp.initDataUnsafe > chat_id with php not js

**Comment by revenkroz** [comment, marked as answer]
You can get `chatId` when person starting bot for the first time — just take it from `Update` object. If you missed it you can send xhr-request with `initData` from frontend (web app) to some special endpoint like this:
```js
fetch('/your-special-endpoint-to-recieve-data', {
    method: 'POST',
    body: Telegram.WebApp.initData, // data from telegram
})
``

**Reply by DeAref** [reply]
But this is very **unsafe**.
I want to know who opened the webapp so I can get **his specific information** from the **database** and show him. Is there another solution?

**Reply by DeAref** [reply]
I want to show sensitive information to the user, so I need to be sure that the user opened this webapp and not someone else.

**Reply by revenkroz** [reply]
Telegram did a really great feature to confirm if this action is from specific user or not. You can validate it ([read more](https://core.telegram.org/bots/webapps#validating-data-received-via-the-web-app))

In PHP it will be something like this:

```php
// $data is parsed data from Telegram
public function checkWebAppData(array $data): bool
{
    asort($data);
    $dataCheckString = [];
    foreach ($data as $key => $value) {
        if ($key === 'hash') {
            continue;
        }

        if (is_array($value)) {
            $dataCheckString[] = $key . '=' . json_encode($value);
        } else {
            $dataCheckString[] = $key . '=' . $value;
        }
    }

    $dataCheckString = implode("\n", $dataCheckString);
    $secretKey = hash_hmac('sha256', 'YOUR-TOKEN-CODE-123', 'WebAppData', true);
    $sig = hash_hmac('sha256', $dataCheckString, $secretKey);

    return $sig === $data['hash'];
}
```

Function will return `true` if data is valid. This way you can get user (`$data['user']`) and trust it.

```php
// $initDataFromTelegram is raw data (string) received by your endpoint
parse_str($initDataFromTelegram, $data);

// check if it is valid
if (checkWebAppData($data) {
    // get user 
    $userData = json_decode($data['user'], true);
    // do something with it
    // ....
}
```

**Reply by revenkroz** [reply]
@DeAref if I answered your question, can I close discussion?

**Reply by DeAref** [reply]
I almost understand what you are saying, but there are still questions in my mind
Where do I get `$initDataFromTelegram`?
This is the same `$update = json_decode(file_get_contents('php://input'));` Is?
**Do you have more example for this method?**

**Reply by revenkroz** [reply]
Yes. It's almost the same, but without `json_decode` as we receive data not in json. You should parse raw data with `parse_str`. The data is `Telegram.WebApp.initData` sent from frontend.

**Reply by CrazyTapok-bit** [reply]
The above code should perform its function properly, but for a more reliable check and the ability to have user data in PHP, you can use the [TgWebValid](https://github.com/CrazyTapok-bit/tgWebValid) library, it contains all the necessary checks and gives a 100% guarantee of the result. Moreover, it will return all user data to you, I recommend you try it. Connection takes only one minute
